// @noir-lang/barretenberg/dest/client_proofs/account_proof/account_prover.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountProver = void 0;
const threads_1 = require("threads");
class AccountProver {
    constructor(prover, mock = false) {
        this.prover = prover;
        this.mock = mock;
    }
    static getCircuitSize(proverless = false) {
        return proverless ? 512 : 32 * 1024;
    }
    async computeKey() {
        const worker = this.prover.getWorker();
        await worker.call('account__init_proving_key', this.mock);
    }
    async loadKey(keyBuf) {
        const worker = this.prover.getWorker();
        const keyPtr = await worker.call('bbmalloc', keyBuf.length);
        await worker.transferToHeap((0, threads_1.Transfer)(keyBuf, [keyBuf.buffer]), keyPtr);
        await worker.call('account__init_proving_key_from_buffer', keyPtr);
        await worker.call('bbfree', keyPtr);
    }
    async getKey() {
        const worker = this.prover.getWorker();
        await worker.acquire();
        try {
            const keySize = await worker.call('account__get_new_proving_key_data', 0);
            const keyPtr = Buffer.from(await worker.sliceMemory(0, 4)).readUInt32LE(0);
            const buf = Buffer.from(await worker.sliceMemory(keyPtr, keyPtr + keySize));
            await worker.call('bbfree', keyPtr);
            return buf;
        }
        finally {
            await worker.release();
        }
    }
    async computeSigningData(tx) {
        const worker = this.prover.getWorker();
        await worker.transferToHeap(tx.toBuffer(), 0);
        await worker.call('account__compute_signing_data', 0, 0);
        return Buffer.from(await worker.sliceMemory(0, 32));
    }
    async createAccountProof(tx, signature) {
        const worker = this.prover.getWorker();
        const buf = Buffer.concat([tx.toBuffer(), signature.toBuffer()]);
        const mem = await worker.call('bbmalloc', buf.length);
        await worker.transferToHeap(buf, mem);
        const proverPtr = await worker.call('account__new_prover', mem, this.mock);
        await worker.call('bbfree', mem);
        const proof = await this.prover.createProof(proverPtr);
        await worker.call('account__delete_prover', proverPtr);
        return proof;
    }
    getProver() {
        return this.prover;
    }
}
exports.AccountProver = AccountProver;
