// @noir-lang/barretenberg/dest/blockchain/blockchain_status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockchainStatusFromJson = exports.blockchainStatusToJson = exports.blockchainBridgeFromJson = exports.blockchainBridgeToJson = exports.blockchainAssetFromJson = exports.blockchainAssetToJson = exports.isAccountTx = exports.isDefiDepositTx = exports.numTxTypes = exports.TxType = void 0;
const address_1 = require("../address");
// TODO: Move to TxType module.
var TxType;
(function (TxType) {
    TxType[TxType["DEPOSIT"] = 0] = "DEPOSIT";
    TxType[TxType["TRANSFER"] = 1] = "TRANSFER";
    TxType[TxType["WITHDRAW_TO_WALLET"] = 2] = "WITHDRAW_TO_WALLET";
    TxType[TxType["WITHDRAW_HIGH_GAS"] = 3] = "WITHDRAW_HIGH_GAS";
    TxType[TxType["ACCOUNT"] = 4] = "ACCOUNT";
    TxType[TxType["DEFI_DEPOSIT"] = 5] = "DEFI_DEPOSIT";
    TxType[TxType["DEFI_CLAIM"] = 6] = "DEFI_CLAIM";
})(TxType = exports.TxType || (exports.TxType = {}));
exports.numTxTypes = 7;
function isDefiDepositTx(txType) {
    return txType === TxType.DEFI_DEPOSIT;
}
exports.isDefiDepositTx = isDefiDepositTx;
function isAccountTx(txType) {
    return txType === TxType.ACCOUNT;
}
exports.isAccountTx = isAccountTx;
const blockchainAssetToJson = ({ address, ...asset }) => ({
    ...asset,
    address: address.toString(),
});
exports.blockchainAssetToJson = blockchainAssetToJson;
const blockchainAssetFromJson = ({ address, ...asset }) => ({
    ...asset,
    address: address_1.EthAddress.fromString(address),
});
exports.blockchainAssetFromJson = blockchainAssetFromJson;
const blockchainBridgeToJson = ({ address, ...bridge }) => ({
    ...bridge,
    address: address.toString(),
});
exports.blockchainBridgeToJson = blockchainBridgeToJson;
const blockchainBridgeFromJson = ({ address, ...bridge }) => ({
    ...bridge,
    address: address_1.EthAddress.fromString(address),
});
exports.blockchainBridgeFromJson = blockchainBridgeFromJson;
function blockchainStatusToJson(status) {
    return {
        ...status,
        rollupContractAddress: status.rollupContractAddress.toString(),
        verifierContractAddress: status.verifierContractAddress.toString(),
        dataRoot: status.dataRoot.toString('hex'),
        nullRoot: status.nullRoot.toString('hex'),
        rootRoot: status.rootRoot.toString('hex'),
        defiRoot: status.defiRoot.toString('hex'),
        defiInteractionHashes: status.defiInteractionHashes.map(v => v.toString('hex')),
        assets: status.assets.map(exports.blockchainAssetToJson),
        bridges: status.bridges.map(exports.blockchainBridgeToJson),
    };
}
exports.blockchainStatusToJson = blockchainStatusToJson;
function blockchainStatusFromJson(json) {
    return {
        ...json,
        rollupContractAddress: address_1.EthAddress.fromString(json.rollupContractAddress),
        verifierContractAddress: address_1.EthAddress.fromString(json.verifierContractAddress),
        dataRoot: Buffer.from(json.dataRoot, 'hex'),
        nullRoot: Buffer.from(json.nullRoot, 'hex'),
        rootRoot: Buffer.from(json.rootRoot, 'hex'),
        defiRoot: Buffer.from(json.defiRoot, 'hex'),
        defiInteractionHashes: json.defiInteractionHashes.map(f => Buffer.from(f, 'hex')),
        assets: json.assets.map(exports.blockchainAssetFromJson),
        bridges: json.bridges.map(exports.blockchainBridgeFromJson),
    };
}
exports.blockchainStatusFromJson = blockchainStatusFromJson;
