// @noir-lang/barretenberg/dest/fft/pooled_fft.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PooledFftFactory = exports.PooledFft = void 0;
const log_1 = require("../log");
const fifo_1 = require("../fifo");
const single_fft_1 = require("./single_fft");
const debug = (0, log_1.createDebugLogger)('bb:fft');
class PooledFft {
    constructor(pool) {
        this.queue = new fifo_1.MemoryFifo();
        this.ffts = pool.workers.map(w => new single_fft_1.SingleFft(w));
    }
    async init(circuitSize) {
        const start = new Date().getTime();
        debug(`initializing fft of size: ${circuitSize}`);
        await Promise.all(this.ffts.map(f => f.init(circuitSize)));
        this.ffts.forEach(w => this.processJobs(w));
        debug(`initialization took: ${new Date().getTime() - start}ms`);
    }
    async destroy() {
        this.queue.cancel();
        await Promise.all(this.ffts.map(f => f.destroy()));
    }
    async processJobs(worker) {
        while (true) {
            const job = await this.queue.get();
            if (!job) {
                break;
            }
            const result = await (job.inverse ? worker.ifft(job.coefficients) : worker.fft(job.coefficients, job.constant));
            job.resolve(result);
        }
    }
    async fft(coefficients, constant) {
        return await new Promise(resolve => this.queue.put({ coefficients, constant, inverse: false, resolve }));
    }
    async ifft(coefficients) {
        return await new Promise(resolve => this.queue.put({ coefficients, inverse: true, resolve }));
    }
}
exports.PooledFft = PooledFft;
class PooledFftFactory {
    constructor(workerPool) {
        this.workerPool = workerPool;
        this.ffts = {};
    }
    async createFft(circuitSize) {
        if (!this.ffts[circuitSize]) {
            const fft = new PooledFft(this.workerPool);
            await fft.init(circuitSize);
            this.ffts[circuitSize] = fft;
        }
        return this.ffts[circuitSize];
    }
    async destroy() {
        await Promise.all(Object.values(this.ffts).map(fft => fft.destroy()));
    }
}
exports.PooledFftFactory = PooledFftFactory;
