// @noir-lang/barretenberg/dest/note_algorithms/defi_interaction_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.packInteractionNotes = exports.computeInteractionHashes = exports.DefiInteractionNote = void 0;
const crypto_1 = require("crypto");
const bigint_buffer_1 = require("../bigint_buffer");
const bridge_id_1 = require("../bridge_id");
const crypto_2 = require("../crypto");
const serialize_1 = require("../serialize");
class DefiInteractionNote {
    constructor(bridgeId, nonce, totalInputValue, totalOutputValueA, totalOutputValueB, result) {
        this.bridgeId = bridgeId;
        this.nonce = nonce;
        this.totalInputValue = totalInputValue;
        this.totalOutputValueA = totalOutputValueA;
        this.totalOutputValueB = totalOutputValueB;
        this.result = result;
    }
    static deserialize(buffer, offset) {
        const des = new serialize_1.Deserializer(buffer, offset);
        const bridgeIdBuffer = des.buffer(32);
        const bridgeId = bridge_id_1.BridgeId.fromBuffer(bridgeIdBuffer);
        const totalInputValue = des.bigInt();
        const totalOutputValueA = des.bigInt();
        const totalOutputValueB = des.bigInt();
        const nonce = des.uInt32();
        const result = des.bool();
        return {
            elem: new DefiInteractionNote(bridgeId, nonce, totalInputValue, totalOutputValueA, totalOutputValueB, result),
            adv: des.getOffset() - offset,
        };
    }
    static random() {
        return new DefiInteractionNote(bridge_id_1.BridgeId.random(), (0, crypto_2.randomBytes)(4).readUInt32BE(0), (0, bigint_buffer_1.toBigIntBE)((0, crypto_2.randomBytes)(32)), (0, bigint_buffer_1.toBigIntBE)((0, crypto_2.randomBytes)(32)), (0, bigint_buffer_1.toBigIntBE)((0, crypto_2.randomBytes)(32)), !!Math.round(Math.random()));
    }
    static fromBuffer(buf) {
        return DefiInteractionNote.deserialize(buf, 0).elem;
    }
    toBuffer() {
        const serializer = new serialize_1.Serializer();
        serializer.buffer(this.bridgeId.toBuffer());
        serializer.bigInt(this.totalInputValue);
        serializer.bigInt(this.totalOutputValueA);
        serializer.bigInt(this.totalOutputValueB);
        serializer.uInt32(this.nonce);
        serializer.bool(this.result);
        return serializer.getBuffer();
    }
    equals(note) {
        return this.toBuffer().equals(note.toBuffer());
    }
}
exports.DefiInteractionNote = DefiInteractionNote;
DefiInteractionNote.EMPTY = new DefiInteractionNote(bridge_id_1.BridgeId.ZERO, 0, BigInt(0), BigInt(0), BigInt(0), false);
DefiInteractionNote.groupModulus = BigInt('21888242871839275222246405745257275088548364400416034343698204186575808495617');
const computeInteractionHashes = (notes, padTo = notes.length) => {
    notes = [...notes, ...Array(padTo - notes.length).fill(DefiInteractionNote.EMPTY)];
    const hash = notes.map(note => (0, crypto_1.createHash)('sha256')
        .update(Buffer.concat([
        note.bridgeId.toBuffer(),
        (0, serialize_1.numToUInt32BE)(note.nonce, 32),
        (0, bigint_buffer_1.toBufferBE)(note.totalInputValue, 32),
        (0, bigint_buffer_1.toBufferBE)(note.totalOutputValueA, 32),
        (0, bigint_buffer_1.toBufferBE)(note.totalOutputValueB, 32),
        Buffer.alloc(31),
        Buffer.from([+note.result]),
    ]))
        .digest());
    return hash.map(h => (0, bigint_buffer_1.toBufferBE)(BigInt('0x' + h.toString('hex')) % DefiInteractionNote.groupModulus, 32));
};
exports.computeInteractionHashes = computeInteractionHashes;
const packInteractionNotes = (notes, padTo = notes.length) => {
    const hash = (0, crypto_1.createHash)('sha256')
        .update(Buffer.concat((0, exports.computeInteractionHashes)(notes, padTo)))
        .digest();
    return (0, bigint_buffer_1.toBufferBE)(BigInt('0x' + hash.toString('hex')) % DefiInteractionNote.groupModulus, 32);
};
exports.packInteractionNotes = packInteractionNotes;
