// @noir-lang/barretenberg/dest/note_algorithms/note_algorithms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoteAlgorithms = void 0;
const bigint_buffer_1 = require("../bigint_buffer");
class NoteAlgorithms {
    constructor(wasm) {
        this.wasm = wasm;
    }
    valueNoteNullifier(noteCommitment, accountPrivateKey, real = true) {
        this.wasm.transferToHeap(noteCommitment, 0);
        this.wasm.transferToHeap(accountPrivateKey, 64);
        this.wasm.call('notes__value_note_nullifier', 0, 64, real, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    valueNoteNullifierBigInt(noteCommitment, accountPrivateKey, real = true) {
        return (0, bigint_buffer_1.toBigIntBE)(this.valueNoteNullifier(noteCommitment, accountPrivateKey, real));
    }
    valueNoteCommitment(note) {
        const noteBuf = note.toBuffer();
        const mem = this.wasm.call('bbmalloc', noteBuf.length);
        this.wasm.transferToHeap(noteBuf, mem);
        this.wasm.call('notes__value_note_commitment', mem, 0);
        this.wasm.call('bbfree', mem);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    valueNotePartialCommitment(noteSecret, owner, accountRequired) {
        this.wasm.transferToHeap(noteSecret, 0);
        this.wasm.transferToHeap(owner.toBuffer(), 32);
        // Currently this is only used for creating the value notes from a claim note.
        // Given these notes are owned by the creator of the claim note, we can leave creator pubkey as 0.
        this.wasm.transferToHeap(Buffer.alloc(32), 96);
        this.wasm.call('notes__value_note_partial_commitment', 0, 32, 96, accountRequired, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    claimNotePartialCommitment(note) {
        const noteBuf = note.toBuffer();
        const mem = this.wasm.call('bbmalloc', noteBuf.length);
        this.wasm.transferToHeap(noteBuf, mem);
        this.wasm.call('notes__claim_note_partial_commitment', mem, 0);
        this.wasm.call('bbfree', mem);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    claimNoteCompletePartialCommitment(partialNote, interactionNonce, fee) {
        this.wasm.transferToHeap(partialNote, 0);
        this.wasm.transferToHeap((0, bigint_buffer_1.toBufferBE)(fee, 32), 32);
        this.wasm.call('notes__claim_note_complete_partial_commitment', 0, interactionNonce, 32, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    claimNoteCommitment(note) {
        const partial = this.claimNotePartialCommitment(note);
        return this.claimNoteCompletePartialCommitment(partial, note.defiInteractionNonce, note.fee);
    }
    claimNoteNullifier(noteCommitment) {
        this.wasm.transferToHeap(noteCommitment, 0);
        this.wasm.call('notes__claim_note_nullifier', 0, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    defiInteractionNoteCommitment(note) {
        const noteBuf = note.toBuffer();
        const mem = this.wasm.call('bbmalloc', noteBuf.length);
        this.wasm.transferToHeap(noteBuf, mem);
        this.wasm.call('notes__defi_interaction_note_commitment', mem, 0);
        this.wasm.call('bbfree', mem);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    accountNoteCommitment(aliasHash, accountPublicKey, spendingPublicKey) {
        this.wasm.transferToHeap(aliasHash.toBuffer32(), 0);
        this.wasm.transferToHeap(accountPublicKey.toBuffer(), 32);
        this.wasm.transferToHeap(spendingPublicKey, 64);
        this.wasm.call('notes__account_note_commitment', 0, 32, 64, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    accountAliasHashNullifier(aliasHash) {
        this.wasm.transferToHeap(aliasHash.toBuffer32(), 0);
        this.wasm.call('notes__compute_account_alias_hash_nullifier', 0, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
    accountPublicKeyNullifier(accountPublicKey) {
        this.wasm.transferToHeap(accountPublicKey.toBuffer(), 0);
        this.wasm.call('notes__compute_account_public_key_nullifier', 0, 0);
        return Buffer.from(this.wasm.sliceMemory(0, 32));
    }
}
exports.NoteAlgorithms = NoteAlgorithms;
