// @noir-lang/barretenberg/dest/note_algorithms/tree_note.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeNote = void 0;
const address_1 = require("../address");
const bigint_buffer_1 = require("../bigint_buffer");
const serialize_1 = require("../serialize");
const viewing_key_1 = require("../viewing_key");
const derive_note_secret_1 = require("./derive_note_secret");
class TreeNote {
    constructor(ownerPubKey, value, assetId, accountRequired, noteSecret, creatorPubKey, inputNullifier) {
        this.ownerPubKey = ownerPubKey;
        this.value = value;
        this.assetId = assetId;
        this.accountRequired = accountRequired;
        this.noteSecret = noteSecret;
        this.creatorPubKey = creatorPubKey;
        this.inputNullifier = inputNullifier;
    }
    toBuffer() {
        return Buffer.concat([
            (0, bigint_buffer_1.toBufferBE)(this.value, 32),
            (0, serialize_1.numToUInt32BE)(this.assetId),
            Buffer.from([this.accountRequired ? 1 : 0]),
            this.ownerPubKey.toBuffer(),
            this.noteSecret,
            this.creatorPubKey,
            this.inputNullifier,
        ]);
    }
    createViewingKey(ephPrivKey, grumpkin) {
        const noteBuf = Buffer.concat([
            (0, bigint_buffer_1.toBufferBE)(this.value, 32),
            (0, serialize_1.numToUInt32BE)(this.assetId),
            (0, serialize_1.numToUInt32BE)(+this.accountRequired),
            this.creatorPubKey,
        ]);
        return viewing_key_1.ViewingKey.createFromEphPriv(noteBuf, this.ownerPubKey, ephPrivKey, grumpkin);
    }
    static fromBuffer(buf) {
        let dataStart = 0;
        const value = (0, bigint_buffer_1.toBigIntBE)(buf.slice(dataStart, dataStart + 32));
        dataStart += 32;
        const assetId = buf.readUInt32BE(dataStart);
        dataStart += 4;
        const accountRequired = !!buf[dataStart];
        dataStart += 1;
        const ownerPubKey = new address_1.GrumpkinAddress(buf.slice(dataStart, dataStart + 64));
        dataStart += 64;
        const noteSecret = buf.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const creatorPubKey = buf.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const inputNullifier = buf.slice(dataStart, dataStart + 32);
        return new TreeNote(ownerPubKey, value, assetId, accountRequired, noteSecret, creatorPubKey, inputNullifier);
    }
    /**
     * Note on how the noteSecret can be derived in two different ways (from ephPubKey or ephPrivKey):
     *
     * ownerPubKey := [ownerPrivKey] * G  (where G is a generator of the grumpkin curve, and `[scalar] * Point` is scalar multiplication).
     *                      ↑
     *         a.k.a. account private key
     *
     * ephPubKey := [ephPrivKey] * G    (where ephPrivKey is a random field element).
     *
     * sharedSecret := [ephPrivKey] * ownerPubKey = [ephPrivKey] * ([ownerPrivKey] * G) = [ownerPrivKey] * ([ephPrivKey] * G) = [ownerPrivKey] * ephPubKey
     *                  ^^^^^^^^^^                                                                                                               ^^^^^^^^^
     * noteSecret is then derivable from the sharedSecret.
     */
    static createFromEphPriv(ownerPubKey, value, assetId, accountRequired, inputNullifier, ephPrivKey, grumpkin, creatorPubKey = Buffer.alloc(32)) {
        const noteSecret = (0, derive_note_secret_1.deriveNoteSecret)(ownerPubKey, ephPrivKey, grumpkin);
        return new TreeNote(ownerPubKey, value, assetId, accountRequired, noteSecret, creatorPubKey, inputNullifier);
    }
    static createFromEphPub(ownerPubKey, value, assetId, accountRequired, inputNullifier, ephPubKey, ownerPrivKey, grumpkin, creatorPubKey = Buffer.alloc(32)) {
        const noteSecret = (0, derive_note_secret_1.deriveNoteSecret)(ephPubKey, ownerPrivKey, grumpkin);
        return new TreeNote(ownerPubKey, value, assetId, accountRequired, noteSecret, creatorPubKey, inputNullifier);
    }
    static recover({ noteBuf, noteSecret }, inputNullifier, ownerPubKey) {
        const value = (0, bigint_buffer_1.toBigIntBE)(noteBuf.slice(0, 32));
        const assetId = noteBuf.readUInt32BE(32);
        const accountRequired = !!noteBuf.readUInt32BE(36);
        const creatorPubKey = noteBuf.slice(40, 72);
        return new TreeNote(ownerPubKey, value, assetId, accountRequired, noteSecret, creatorPubKey, inputNullifier);
    }
}
exports.TreeNote = TreeNote;
TreeNote.EMPTY = new TreeNote(address_1.GrumpkinAddress.one(), BigInt(0), 0, false, Buffer.alloc(32), Buffer.alloc(32), Buffer.alloc(32));
TreeNote.SIZE = TreeNote.EMPTY.toBuffer().length;
