// @noir-lang/barretenberg/dest/block_source/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Block = void 0;
const tslib_1 = require("tslib");
const serialize_1 = require("../serialize");
const blockchain_1 = require("../blockchain");
const defi_interaction_event_1 = require("./defi_interaction_event");
class Block {
    constructor(txHash, created, rollupId, rollupSize, encodedRollupProofData, offchainTxData, interactionResult, gasUsed, gasPrice, subtreeRoot) {
        this.txHash = txHash;
        this.created = created;
        this.rollupId = rollupId;
        this.rollupSize = rollupSize;
        this.encodedRollupProofData = encodedRollupProofData;
        this.offchainTxData = offchainTxData;
        this.interactionResult = interactionResult;
        this.gasUsed = gasUsed;
        this.gasPrice = gasPrice;
        this.subtreeRoot = subtreeRoot;
    }
    static deserialize(buf, offset = 0) {
        const des = new serialize_1.Deserializer(buf, offset);
        const txHash = des.exec(blockchain_1.TxHash.deserialize);
        const created = des.date();
        const rollupId = des.uInt32();
        const rollupSize = des.uInt32();
        const rollupProofData = des.vector();
        const offchainTxData = des.deserializeArray(serialize_1.deserializeBufferFromVector);
        const interactionResult = des.deserializeArray(defi_interaction_event_1.DefiInteractionEvent.deserialize);
        const gasUsed = des.uInt32();
        const gasPrice = des.bigInt();
        const subtreeRoot = des.vector();
        return {
            elem: new Block(txHash, created, rollupId, rollupSize, rollupProofData, offchainTxData, interactionResult, gasUsed, gasPrice, subtreeRoot.equals(Buffer.alloc(0)) ? undefined : subtreeRoot),
            adv: des.getOffset() - offset,
        };
    }
    static fromBuffer(buf) {
        return Block.deserialize(buf).elem;
    }
    toBuffer() {
        var _a;
        return Buffer.concat([
            this.txHash.toBuffer(),
            (0, serialize_1.serializeDate)(this.created),
            (0, serialize_1.numToUInt32BE)(this.rollupId),
            (0, serialize_1.numToUInt32BE)(this.rollupSize),
            (0, serialize_1.serializeBufferToVector)(this.encodedRollupProofData),
            (0, serialize_1.serializeBufferArrayToVector)(this.offchainTxData.map(b => (0, serialize_1.serializeBufferToVector)(b))),
            (0, serialize_1.serializeBufferArrayToVector)(this.interactionResult.map(b => b.toBuffer())),
            (0, serialize_1.numToUInt32BE)(this.gasUsed),
            (0, serialize_1.serializeBigInt)(this.gasPrice),
            (0, serialize_1.serializeBufferToVector)((_a = this.subtreeRoot) !== null && _a !== void 0 ? _a : Buffer.alloc(0)),
        ]);
    }
}
exports.Block = Block;
(0, tslib_1.__exportStar)(require("./server_block_source"), exports);
(0, tslib_1.__exportStar)(require("./defi_interaction_event"), exports);
