// @noir-lang/barretenberg/dest/block_source/server_block_source.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerBlockSource = void 0;
const _1 = require(".");
const events_1 = require("events");
const iso_fetch_1 = require("../iso_fetch");
const serialize_1 = require("../serialize");
// import { createDebugLogger } from '@aztec/barretenberg/log';
// const debug = createDebugLogger('bb:server_block_source');
class ServerBlockSource extends events_1.EventEmitter {
    constructor(baseUrl, pollInterval = 10000) {
        super();
        this.pollInterval = pollInterval;
        this.running = false;
        this.runningPromise = Promise.resolve();
        this.interruptPromise = Promise.resolve();
        this.interruptResolve = () => { };
        this.baseUrl = baseUrl.toString().replace(/\/$/, '');
    }
    async getLatestRollupId() {
        const url = new URL(`${this.baseUrl}/get-blocks`);
        const response = await this.awaitSucceed(() => (0, iso_fetch_1.fetch)(url.toString(), { headers: { 'Accept-encoding': 'gzip' } }));
        const result = Buffer.from(await response.arrayBuffer());
        const des = new serialize_1.Deserializer(result);
        return des.int32();
    }
    async start(from = 0) {
        this.running = true;
        this.interruptPromise = new Promise(resolve => (this.interruptResolve = resolve));
        const emitBlocks = async () => {
            try {
                const blocks = await this.getBlocks(from);
                for (const block of blocks) {
                    this.emit('block', block);
                    from = block.rollupId + 1;
                }
            }
            catch (err) {
                // debug(err);
            }
        };
        await emitBlocks();
        const poll = async () => {
            while (this.running) {
                await emitBlocks();
                await this.sleepOrInterrupted(this.pollInterval);
            }
        };
        this.runningPromise = poll();
    }
    stop() {
        this.running = false;
        this.interruptResolve();
        return this.runningPromise;
    }
    async awaitSucceed(fn) {
        while (true) {
            try {
                const response = await fn();
                if (response.status !== 200) {
                    throw new Error(`Bad status code: ${response.status}`);
                }
                return response;
            }
            catch (err) {
                console.log(err.message);
                await new Promise(resolve => setTimeout(resolve, 10000));
            }
        }
    }
    async getBlocks(from) {
        const url = new URL(`${this.baseUrl}/get-blocks`);
        if (from !== undefined) {
            url.searchParams.append('from', from.toString());
        }
        const response = await this.awaitSucceed(() => (0, iso_fetch_1.fetch)(url.toString(), { headers: { 'Accept-encoding': 'gzip' } }));
        const result = Buffer.from(await response.arrayBuffer());
        const des = new serialize_1.Deserializer(result);
        des.int32();
        return des.deserializeArray(_1.Block.deserialize);
    }
    async sleepOrInterrupted(ms) {
        let timeout;
        const promise = new Promise(resolve => (timeout = setTimeout(resolve, ms)));
        await Promise.race([promise, this.interruptPromise]);
        clearTimeout(timeout);
    }
}
exports.ServerBlockSource = ServerBlockSource;
