// @noir-lang/barretenberg/dest/serialize/deserializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deserializer = void 0;
const free_funcs_1 = require("./free_funcs");
class Deserializer {
    constructor(buf, offset = 0) {
        this.buf = buf;
        this.offset = offset;
    }
    bool() {
        return this.exec(free_funcs_1.deserializeBool) ? true : false;
    }
    uInt32() {
        return this.exec(free_funcs_1.deserializeUInt32);
    }
    int32() {
        return this.exec(free_funcs_1.deserializeInt32);
    }
    bigInt(width = 32) {
        return this.exec((buf, offset) => (0, free_funcs_1.deserializeBigInt)(buf, offset, width));
    }
    vector() {
        return this.exec(free_funcs_1.deserializeBufferFromVector);
    }
    buffer(width) {
        const buf = this.buf.slice(this.offset, this.offset + width);
        this.offset += width;
        return buf;
    }
    string() {
        return this.vector().toString();
    }
    date() {
        return new Date(Number(this.bigInt(8)));
    }
    deserializeArray(fn) {
        return this.exec((buf, offset) => (0, free_funcs_1.deserializeArrayFromVector)(fn, buf, offset));
    }
    exec(fn) {
        const { elem, adv } = fn(this.buf, this.offset);
        this.offset += adv;
        return elem;
    }
    getOffset() {
        return this.offset;
    }
}
exports.Deserializer = Deserializer;
