// @noir-lang/barretenberg/dest/client_proofs/generic_proof/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verify_proof = exports.create_proof_with_witness = exports.create_proof = exports.setup_generic_prover_and_verifier = void 0;
const tslib_1 = require("tslib");
// TODO: fix this typo of barretenberg spelling
const aztec_backend_1 = require("@noir-lang/aztec_backend");
const crs_1 = require("../../crs");
const fft_1 = require("../../fft");
const pippenger_1 = require("../../pippenger");
const wasm_1 = require("../../wasm");
const prover_1 = require("../prover");
const standard_example_prover_1 = require("./standard_example_prover");
const standard_example_verifier_1 = require("./standard_example_verifier");
(0, tslib_1.__exportStar)(require("./standard_example_prover"), exports);
(0, tslib_1.__exportStar)(require("./standard_example_verifier"), exports);
// Takes in a structure representing ACIR which we get from
// wasm version of the noir compiler
//
// We then serialise the ACIR into a byte array which represents
// what is known as standard_format in the C++ code.
//
// standard_format is capable of representing any statement.
//
// Note this byte array would be analogous to the `tx` structures
// in the other client proofs
async function setup_generic_prover_and_verifier(acir) {
    const serialised_circuit = (0, aztec_backend_1.serialise_acir_to_barrtenberg_circuit)(acir);
    const barretenberg = await wasm_1.BarretenbergWasm.new();
    const circSize = await (0, standard_example_prover_1.getCircuitSize)(barretenberg, serialised_circuit);
    const crs = await load_crs(circSize);
    const numWorkers = getNumCores();
    const wasm = await wasm_1.BarretenbergWasm.new();
    const workerPool = await wasm_1.WorkerPool.new(wasm, numWorkers);
    const pippenger = new pippenger_1.PooledPippenger(workerPool);
    const fft = new fft_1.PooledFft(workerPool);
    await fft.init(circSize);
    await pippenger.init(crs.getData());
    const prover = new prover_1.Prover(workerPool.workers[0], pippenger, fft);
    const standardExampleProver = new standard_example_prover_1.StandardExampleProver(prover);
    await standardExampleProver.initCircuitDefinition(serialised_circuit);
    const standardExampleVerifier = new standard_example_verifier_1.StandardExampleVerifier();
    // Create proving key with a dummy CRS
    await standardExampleProver.computeKey();
    // Create verifier key *and* patch proving key with the CRS
    await standardExampleVerifier.computeKey(pippenger.pool[0], crs.getG2Data());
    // Compute smart contract and cache it
    await standardExampleVerifier.computeSmartContract(pippenger.pool[0], crs.getG2Data(), serialised_circuit);
    return Promise.all([standardExampleProver, standardExampleVerifier]);
}
exports.setup_generic_prover_and_verifier = setup_generic_prover_and_verifier;
async function load_crs(circSize) {
    // We may need more elements in the SRS than the circuit size. In particular, we may need circSize +1
    // We add an offset here to account for that
    const offset = 1;
    const crs = new crs_1.Crs(circSize + offset);
    await crs.download();
    return crs;
}
async function create_proof(prover, acir, abi) {
    // compute partial witness here
    let witness_arr = await compute_partial_witnesses(acir, abi);
    // computes the proof
    const proof = await prover.createProof(witness_arr);
    return proof;
}
exports.create_proof = create_proof;
async function create_proof_with_witness(prover, witness_arr) {
    // computes the proof
    const proof = await prover.createProof(witness_arr);
    return proof;
}
exports.create_proof_with_witness = create_proof_with_witness;
async function verify_proof(verifier, proof) {
    const verified = await verifier.verifyProof(proof);
    return verified;
}
exports.verify_proof = verify_proof;
async function compute_partial_witnesses(circuit, abi) {
    // Use the ACIR representation to compute the partial witnesses
    // Assumption: .values() will always return the values in a deterministic order;
    // (from left to right) in the abi object
    let values = [];
    for (const [_, v] of Object.entries(abi)) {
        let entry_values = AnyToHexStrs(v);
        values = values.concat(entry_values);
    }
    return (0, aztec_backend_1.compute_witnesses)(circuit, values);
}
function AnyToHexStrs(any_object) {
    let values = [];
    if (Array.isArray(any_object)) {
        for (let variable of any_object) {
            values = values.concat(AnyToHexStrs(variable));
        }
    }
    else if (typeof any_object === 'string' || any_object instanceof String) {
        // If the type is a string, we expect it to be a hex string
        let string_object = any_object;
        if (isValidHex(string_object)) {
            values.push(string_object);
        }
        else {
            // TODO: throw should not be in a library, but currently we aren't doing 
            // TODO: much in terms of error handling
            throw new Error("strings can only be hexadecimal and must start with 0x");
        }
    }
    else if (Number.isInteger(any_object)) {
        let number_object = any_object;
        let number_hex = number_object.toString(16);
        // The rust code only accepts even hex digits
        let is_even_hex_length = number_hex.length % 2 == 0;
        if (is_even_hex_length) {
            values.push("0x" + number_hex);
        }
        else {
            values.push("0x0" + number_hex);
        }
    }
    else {
        throw new Error("unknown object type in the abi");
    }
    return values;
}
function isValidHex(hex_str) {
    return !isNaN(Number(hex_str));
}
function getNumCores() {
    // TODO: The below comment was when we had this in a separate package
    //
    // Barretenberg.js uses navigator.hardwareConcurrency which is
    // only available in the desktop environment, not in js
    //
    // No need to find a polyfill for it, as our circuit is so small
    return 4;
}
