// @noir-lang/barretenberg/dest/wasm/barretenberg_wasm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarretenbergWasm = exports.fetchCode = void 0;
const tslib_1 = require("tslib");
const fs_1 = require("fs");
const detect_node_1 = (0, tslib_1.__importDefault)(require("detect-node"));
const util_1 = require("util");
const events_1 = require("events");
const log_1 = require("../log");
const crypto_1 = require("../crypto");
const fifo_1 = require("../fifo");
events_1.EventEmitter.defaultMaxListeners = 30;
async function fetchCode() {
    if (detect_node_1.default) {
        return await (0, util_1.promisify)(fs_1.readFile)(__dirname + '/barretenberg.wasm');
    }
    else {
        const res = await fetch('/barretenberg.wasm');
        return Buffer.from(await res.arrayBuffer());
    }
}
exports.fetchCode = fetchCode;
class BarretenbergWasm extends events_1.EventEmitter {
    constructor() {
        super();
        this.mutexQ = new fifo_1.MemoryFifo();
        this.mutexQ.put(true);
    }
    static async new(name = 'wasm', initial) {
        const barretenberg = new BarretenbergWasm();
        barretenberg.on('log', (0, log_1.createDebugLogger)(`bb:${name}`));
        await barretenberg.init(undefined, initial);
        return barretenberg;
    }
    async init(module, initial = 256) {
        this.emit('log', `intial mem: ${initial}`);
        this.memory = new WebAssembly.Memory({ initial, maximum: 65536 });
        this.heap = new Uint8Array(this.memory.buffer);
        const importObj = {
            /* eslint-disable camelcase */
            wasi_snapshot_preview1: {
                environ_get: () => { },
                environ_sizes_get: () => { },
                fd_close: () => { },
                fd_read: () => { },
                fd_write: () => { },
                fd_seek: () => { },
                fd_fdstat_get: () => { },
                fd_fdstat_set_flags: () => { },
                path_open: () => { },
                path_filestat_get: () => { },
                proc_exit: () => { },
                random_get: (arr, length) => {
                    arr = arr >>> 0;
                    const heap = new Uint8Array(this.memory.buffer);
                    const randomData = (0, crypto_1.randomBytes)(length);
                    for (let i = arr; i < arr + length; ++i) {
                        heap[i] = randomData[i - arr];
                    }
                },
            },
            /* eslint-enable camelcase */
            module: {},
            env: {
                logstr: (addr) => {
                    addr = addr >>> 0;
                    const m = this.getMemory();
                    let i = addr;
                    for (; m[i] !== 0; ++i)
                        ;
                    // eslint-disable-next-line
                    const decoder = detect_node_1.default ? new (require('util').TextDecoder)() : new TextDecoder();
                    const str = decoder.decode(m.slice(addr, i));
                    const str2 = `${str} (mem:${m.length})`;
                    this.emit('log', str2);
                },
                memory: this.memory,
            },
        };
        if (module) {
            this.instance = await WebAssembly.instantiate(module, importObj);
            this.module = module;
        }
        else {
            const { instance, module } = await WebAssembly.instantiate(await fetchCode(), importObj);
            this.instance = instance;
            this.module = module;
        }
    }
    exports() {
        return this.instance.exports;
    }
    /**
     * When returning values from the WASM, use >>> operator to convert signed representation to unsigned representation.
     */
    call(name, ...args) {
        if (!this.exports()[name]) {
            throw new Error(`WASM function ${name} not found.`);
        }
        try {
            return this.exports()[name](...args) >>> 0;
        }
        catch (err) {
            const message = `WASM function ${name} aborted, error: ${err}`;
            this.emit('log', message);
            throw new Error(message);
        }
    }
    getMemory() {
        if (this.heap.length === 0) {
            return new Uint8Array(this.memory.buffer);
        }
        return this.heap;
    }
    memSize() {
        return this.getMemory().length;
    }
    sliceMemory(start, end) {
        return this.getMemory().slice(start, end);
    }
    transferToHeap(arr, offset) {
        const mem = this.getMemory();
        for (let i = 0; i < arr.length; i++) {
            mem[i + offset] = arr[i];
        }
    }
    /**
     * When calling the wasm, sometimes a caller will require exclusive access over a series of calls.
     * e.g. When a result is written to address 0, one cannot have another caller writing to the same address via
     * transferToHeap before the result is read via sliceMemory.
     * acquire() gets a single token from a fifo. The caller must call release() to add the token back.
     */
    async acquire() {
        await this.mutexQ.get();
    }
    release() {
        if (this.mutexQ.length() !== 0) {
            throw new Error('Release called but not acquired.');
        }
        this.mutexQ.put(true);
    }
}
exports.BarretenbergWasm = BarretenbergWasm;
