// @noir-lang/barretenberg/dest/wasm/worker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable require-await */
const observable_1 = require("threads/observable");
const worker_1 = require("threads/worker");
const _1 = require(".");
let wasm;
const subject = new observable_1.Subject();
const worker = {
    async init(module, initial) {
        wasm = new _1.BarretenbergWasm();
        wasm.on('log', str => subject.next(str));
        await wasm.init(module, initial);
    },
    async transferToHeap(buffer, offset) {
        wasm.transferToHeap(buffer, offset);
    },
    async sliceMemory(start, end) {
        const mem = wasm.sliceMemory(start, end);
        return (0, worker_1.Transfer)(mem, [mem.buffer]);
    },
    async call(name, ...args) {
        return wasm.call(name, ...args);
    },
    async memSize() {
        return wasm.memSize();
    },
    logs() {
        return observable_1.Observable.from(subject);
    },
    /**
     * When calling the wasm, sometimes a caller will require exclusive access over a series of calls.
     * e.g. When a result is written to address 0, one cannot have another caller writing to the same address via
     * transferToHeap before the result is read via sliceMemory.
     * acquire() gets a single token from a fifo. The caller must call release() to add the token back.
     */
    async acquire() {
        await wasm.acquire();
    },
    async release() {
        wasm.release();
    },
};
(0, worker_1.expose)(worker);
