// @noir-lang/barretenberg/dest/offchain_tx_data/offchain_defi_deposit_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OffchainDefiDepositData = void 0;
const address_1 = require("../address");
const bigint_buffer_1 = require("../bigint_buffer");
const bridge_id_1 = require("../bridge_id");
const serialize_1 = require("../serialize");
const viewing_key_1 = require("../viewing_key");
class OffchainDefiDepositData {
    constructor(bridgeId, partialState, partialStateSecretEphPubKey, // the public key from which the partial state's secret may be derived (when combined with a valid account private key).
    depositValue, txFee, viewingKey, // viewing key for the 'change' note
    txRefNo = 0) {
        this.bridgeId = bridgeId;
        this.partialState = partialState;
        this.partialStateSecretEphPubKey = partialStateSecretEphPubKey;
        this.depositValue = depositValue;
        this.txFee = txFee;
        this.viewingKey = viewingKey;
        this.txRefNo = txRefNo;
        if (partialState.length !== 32) {
            throw new Error('Expect partialState to be 32 bytes.');
        }
        if (viewingKey.isEmpty()) {
            throw new Error('Viewing key cannot be empty.');
        }
    }
    static fromBuffer(buf) {
        if (buf.length !== OffchainDefiDepositData.SIZE) {
            throw new Error('Invalid buffer size.');
        }
        let dataStart = 0;
        const bridgeId = bridge_id_1.BridgeId.fromBuffer(buf.slice(dataStart, dataStart + 32));
        dataStart += 32;
        const partialState = buf.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const partialStateSecretEphPubKey = new address_1.GrumpkinAddress(buf.slice(dataStart, dataStart + 64));
        dataStart += 64;
        const depositValue = (0, bigint_buffer_1.toBigIntBE)(buf.slice(dataStart, dataStart + 32));
        dataStart += 32;
        const txFee = (0, bigint_buffer_1.toBigIntBE)(buf.slice(dataStart, dataStart + 32));
        dataStart += 32;
        const viewingKey = new viewing_key_1.ViewingKey(buf.slice(dataStart, dataStart + viewing_key_1.ViewingKey.SIZE));
        dataStart += viewing_key_1.ViewingKey.SIZE;
        const txRefNo = buf.readUInt32BE(dataStart);
        return new OffchainDefiDepositData(bridgeId, partialState, partialStateSecretEphPubKey, depositValue, txFee, viewingKey, txRefNo);
    }
    toBuffer() {
        return Buffer.concat([
            this.bridgeId.toBuffer(),
            this.partialState,
            this.partialStateSecretEphPubKey.toBuffer(),
            (0, bigint_buffer_1.toBufferBE)(this.depositValue, 32),
            (0, bigint_buffer_1.toBufferBE)(this.txFee, 32),
            this.viewingKey.toBuffer(),
            (0, serialize_1.numToUInt32BE)(this.txRefNo),
        ]);
    }
}
exports.OffchainDefiDepositData = OffchainDefiDepositData;
OffchainDefiDepositData.EMPTY = new OffchainDefiDepositData(bridge_id_1.BridgeId.ZERO, Buffer.alloc(32), address_1.GrumpkinAddress.ZERO, BigInt(0), BigInt(0), new viewing_key_1.ViewingKey(Buffer.alloc(viewing_key_1.ViewingKey.SIZE)));
OffchainDefiDepositData.SIZE = OffchainDefiDepositData.EMPTY.toBuffer().length;
