// @noir-lang/barretenberg/dest/offchain_tx_data/offchain_account_data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OffchainAccountData = void 0;
const account_id_1 = require("../account_id");
const address_1 = require("../address");
const serialize_1 = require("../serialize");
class OffchainAccountData {
    constructor(accountPublicKey, aliasHash, spendingPublicKey1 = Buffer.alloc(32), spendingPublicKey2 = Buffer.alloc(32), txRefNo = 0) {
        this.accountPublicKey = accountPublicKey;
        this.aliasHash = aliasHash;
        this.spendingPublicKey1 = spendingPublicKey1;
        this.spendingPublicKey2 = spendingPublicKey2;
        this.txRefNo = txRefNo;
        if (spendingPublicKey1.length !== 32) {
            throw new Error('Expect spendingPublicKey1 to be 32 bytes.');
        }
        if (spendingPublicKey2.length !== 32) {
            throw new Error('Expect spendingPublicKey2 to be 32 bytes.');
        }
    }
    static fromBuffer(buf) {
        if (buf.length !== OffchainAccountData.SIZE) {
            throw new Error(`Invalid buffer size: ${buf.length} != ${OffchainAccountData.SIZE}`);
        }
        let dataStart = 0;
        const accountPublicKey = new address_1.GrumpkinAddress(buf.slice(dataStart, dataStart + 64));
        dataStart += 64;
        const aliasHash = new account_id_1.AliasHash(buf.slice(dataStart, dataStart + account_id_1.AliasHash.SIZE));
        dataStart += account_id_1.AliasHash.SIZE;
        const spendingPublicKey1 = buf.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const spendingPublicKey2 = buf.slice(dataStart, dataStart + 32);
        dataStart += 32;
        const txRefNo = buf.readUInt32BE(dataStart);
        return new OffchainAccountData(accountPublicKey, aliasHash, spendingPublicKey1, spendingPublicKey2, txRefNo);
    }
    toBuffer() {
        return Buffer.concat([
            this.accountPublicKey.toBuffer(),
            this.aliasHash.toBuffer(),
            this.spendingPublicKey1,
            this.spendingPublicKey2,
            (0, serialize_1.numToUInt32BE)(this.txRefNo),
        ]);
    }
}
exports.OffchainAccountData = OffchainAccountData;
OffchainAccountData.EMPTY = new OffchainAccountData(address_1.GrumpkinAddress.ZERO, account_id_1.AliasHash.ZERO);
OffchainAccountData.SIZE = OffchainAccountData.EMPTY.toBuffer().length;
