// @noir-lang/barretenberg/dest/rollup_provider/rollup_provider_status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rollupProviderStatusFromJson = exports.rollupProviderStatusToJson = exports.partialRuntimeConfigFromJson = exports.runtimeConfigFromJson = exports.runtimeConfigToJson = void 0;
const tslib_1 = require("tslib");
const address_1 = require("../address");
const blockchain_1 = require("../blockchain");
const bridge_config_1 = require("./bridge_config");
const bridge_status_1 = require("./bridge_status");
const privacy_set_1 = require("./privacy_set");
(0, tslib_1.__exportStar)(require("./bridge_config"), exports);
(0, tslib_1.__exportStar)(require("./bridge_status"), exports);
(0, tslib_1.__exportStar)(require("./privacy_set"), exports);
const runtimeConfigToJson = ({ maxFeeGasPrice, maxFeePerGas, maxPriorityFeePerGas, bridgeConfigs, privacySets, rollupBeneficiary, ...rest }) => ({
    ...rest,
    maxFeeGasPrice: maxFeeGasPrice.toString(),
    maxFeePerGas: maxFeePerGas.toString(),
    maxPriorityFeePerGas: maxPriorityFeePerGas.toString(),
    bridgeConfigs: bridgeConfigs.map(bridge_config_1.bridgeConfigToJson),
    privacySets: (0, privacy_set_1.privacySetsToJson)(privacySets),
    rollupBeneficiary: rollupBeneficiary ? rollupBeneficiary.toString() : undefined,
});
exports.runtimeConfigToJson = runtimeConfigToJson;
const runtimeConfigFromJson = ({ maxFeeGasPrice, maxFeePerGas, maxPriorityFeePerGas, bridgeConfigs, privacySets, rollupBeneficiary, ...rest }) => ({
    ...rest,
    maxFeeGasPrice: BigInt(maxFeeGasPrice),
    maxFeePerGas: BigInt(maxFeePerGas),
    maxPriorityFeePerGas: BigInt(maxPriorityFeePerGas),
    bridgeConfigs: bridgeConfigs.map(bridge_config_1.bridgeConfigFromJson),
    privacySets: (0, privacy_set_1.privacySetsFromJson)(privacySets),
    rollupBeneficiary: rollupBeneficiary ? address_1.EthAddress.fromString(rollupBeneficiary) : undefined,
});
exports.runtimeConfigFromJson = runtimeConfigFromJson;
const partialRuntimeConfigFromJson = ({ maxFeeGasPrice, maxFeePerGas, maxPriorityFeePerGas, bridgeConfigs, privacySets, rollupBeneficiary, ...rest }) => ({
    ...rest,
    ...(maxFeeGasPrice !== undefined ? { maxFeeGasPrice: BigInt(maxFeeGasPrice) } : {}),
    ...(maxFeePerGas !== undefined ? { maxFeePerGas: BigInt(maxFeePerGas) } : {}),
    ...(maxPriorityFeePerGas !== undefined ? { maxPriorityFeePerGas: BigInt(maxPriorityFeePerGas) } : {}),
    ...(bridgeConfigs ? { bridgeConfigs: bridgeConfigs.map(bridge_config_1.bridgeConfigFromJson) } : {}),
    ...(privacySets ? { privacySets: (0, privacy_set_1.privacySetsFromJson)(privacySets) } : {}),
    ...(rollupBeneficiary ? { rollupBeneficiary: address_1.EthAddress.fromString(rollupBeneficiary) } : {}),
});
exports.partialRuntimeConfigFromJson = partialRuntimeConfigFromJson;
const rollupProviderStatusToJson = ({ blockchainStatus, nextPublishTime, runtimeConfig, bridgeStatus, ...rest }) => ({
    ...rest,
    blockchainStatus: (0, blockchain_1.blockchainStatusToJson)(blockchainStatus),
    nextPublishTime: nextPublishTime.toISOString(),
    runtimeConfig: (0, exports.runtimeConfigToJson)(runtimeConfig),
    bridgeStatus: bridgeStatus.map(bridge_status_1.bridgeStatusToJson),
});
exports.rollupProviderStatusToJson = rollupProviderStatusToJson;
const rollupProviderStatusFromJson = ({ blockchainStatus, nextPublishTime, runtimeConfig, bridgeStatus, ...rest }) => ({
    ...rest,
    blockchainStatus: (0, blockchain_1.blockchainStatusFromJson)(blockchainStatus),
    nextPublishTime: new Date(nextPublishTime),
    runtimeConfig: (0, exports.runtimeConfigFromJson)(runtimeConfig),
    bridgeStatus: bridgeStatus.map(bridge_status_1.bridgeStatusFromJson),
});
exports.rollupProviderStatusFromJson = rollupProviderStatusFromJson;
