// @noir-lang/barretenberg/dest/rollup_provider/rollup_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.depositTxFromJson = exports.depositTxToJson = exports.initialWorldStateFromBuffer = exports.initialWorldStateToBuffer = exports.pendingTxFromJson = exports.pendingTxToJson = exports.txFromJson = exports.txToJson = exports.DefiSettlementTime = exports.TxSettlementTime = void 0;
const address_1 = require("../address");
const tx_id_1 = require("../tx_id");
var TxSettlementTime;
(function (TxSettlementTime) {
    TxSettlementTime[TxSettlementTime["NEXT_ROLLUP"] = 0] = "NEXT_ROLLUP";
    TxSettlementTime[TxSettlementTime["INSTANT"] = 1] = "INSTANT";
})(TxSettlementTime = exports.TxSettlementTime || (exports.TxSettlementTime = {}));
var DefiSettlementTime;
(function (DefiSettlementTime) {
    DefiSettlementTime[DefiSettlementTime["DEADLINE"] = 0] = "DEADLINE";
    DefiSettlementTime[DefiSettlementTime["NEXT_ROLLUP"] = 1] = "NEXT_ROLLUP";
    DefiSettlementTime[DefiSettlementTime["INSTANT"] = 2] = "INSTANT";
})(DefiSettlementTime = exports.DefiSettlementTime || (exports.DefiSettlementTime = {}));
const txToJson = ({ proofData, offchainTxData, depositSignature }) => ({
    proofData: proofData.toString('hex'),
    offchainTxData: offchainTxData.toString('hex'),
    depositSignature: depositSignature ? depositSignature.toString('hex') : undefined,
});
exports.txToJson = txToJson;
const txFromJson = ({ proofData, offchainTxData, depositSignature }) => ({
    proofData: Buffer.from(proofData, 'hex'),
    offchainTxData: Buffer.from(offchainTxData, 'hex'),
    depositSignature: depositSignature ? Buffer.from(depositSignature, 'hex') : undefined,
});
exports.txFromJson = txFromJson;
const pendingTxToJson = ({ txId, noteCommitment1, noteCommitment2 }) => ({
    txId: txId.toString(),
    noteCommitment1: noteCommitment1.toString('hex'),
    noteCommitment2: noteCommitment2.toString('hex'),
});
exports.pendingTxToJson = pendingTxToJson;
const pendingTxFromJson = ({ txId, noteCommitment1, noteCommitment2 }) => ({
    txId: tx_id_1.TxId.fromString(txId),
    noteCommitment1: Buffer.from(noteCommitment1, 'hex'),
    noteCommitment2: Buffer.from(noteCommitment2, 'hex'),
});
exports.pendingTxFromJson = pendingTxFromJson;
const initialWorldStateToBuffer = (initialWorldState) => {
    const accountsSizeBuf = Buffer.alloc(4);
    accountsSizeBuf.writeUInt32BE(initialWorldState.initialAccounts.length);
    return Buffer.concat([accountsSizeBuf, initialWorldState.initialAccounts, ...initialWorldState.initialSubtreeRoots]);
};
exports.initialWorldStateToBuffer = initialWorldStateToBuffer;
const initialWorldStateFromBuffer = (data) => {
    const accountsSize = data.readUInt32BE(0);
    const subTreeStart = 4 + accountsSize;
    const initialWorldState = {
        initialAccounts: data.slice(4, subTreeStart),
        initialSubtreeRoots: [],
    };
    // each sub tree root is 32 bytes
    for (let i = subTreeStart; i < data.length; i += 32) {
        initialWorldState.initialSubtreeRoots.push(data.slice(i, i + 32));
    }
    return initialWorldState;
};
exports.initialWorldStateFromBuffer = initialWorldStateFromBuffer;
const depositTxToJson = ({ assetId, value, publicOwner }) => ({
    assetId,
    value: value.toString(),
    publicOwner: publicOwner.toString(),
});
exports.depositTxToJson = depositTxToJson;
const depositTxFromJson = ({ assetId, value, publicOwner }) => ({
    assetId,
    value: BigInt(value),
    publicOwner: address_1.EthAddress.fromString(publicOwner),
});
exports.depositTxFromJson = depositTxFromJson;
