// @noir-lang/barretenberg/dest/rollup_provider/server_rollup_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerRollupProvider = void 0;
const asset_1 = require("../asset");
const block_source_1 = require("../block_source");
const iso_fetch_1 = require("../iso_fetch");
const tx_id_1 = require("../tx_id");
const rollup_provider_1 = require("./rollup_provider");
const rollup_provider_status_1 = require("./rollup_provider_status");
class ServerRollupProvider extends block_source_1.ServerBlockSource {
    constructor(baseUrl, pollInterval = 10000) {
        super(baseUrl, pollInterval);
    }
    async sendTxs(txs) {
        const data = txs.map(rollup_provider_1.txToJson);
        const response = await this.fetch('/txs', data);
        const body = await response.json();
        return body.txIds.map(txId => tx_id_1.TxId.fromString(txId));
    }
    async getTxFees(assetId) {
        const response = await this.fetch('/tx-fees', { assetId });
        const txFees = (await response.json());
        return txFees.map(fees => fees.map(asset_1.assetValueFromJson));
    }
    async getDefiFees(bridgeId) {
        const response = await this.fetch('/defi-fees', { bridgeId: bridgeId.toString() });
        const defiFees = (await response.json());
        return defiFees.map(asset_1.assetValueFromJson);
    }
    async getStatus() {
        const response = await this.fetch('/status');
        try {
            return (0, rollup_provider_status_1.rollupProviderStatusFromJson)(await response.json());
        }
        catch (err) {
            throw new Error('Bad response: getStatus()');
        }
    }
    async getPendingTxs() {
        const response = await this.fetch('/get-pending-txs');
        const txs = await response.json();
        return txs.map(rollup_provider_1.pendingTxFromJson);
    }
    async getPendingNoteNullifiers() {
        const response = await this.fetch('/get-pending-note-nullifiers');
        const nullifiers = (await response.json());
        return nullifiers.map(n => Buffer.from(n, 'hex'));
    }
    async getPendingDepositTxs() {
        const response = await this.fetch('/get-pending-deposit-txs');
        const txs = await response.json();
        return txs.map(rollup_provider_1.depositTxFromJson);
    }
    async clientLog(log) {
        await this.fetch('/client-log', log);
    }
    async getInitialWorldState() {
        const response = await this.fetch('/get-initial-world-state');
        const arrBuffer = await response.arrayBuffer();
        return (0, rollup_provider_1.initialWorldStateFromBuffer)(Buffer.from(arrBuffer));
    }
    async isAccountRegistered(accountPublicKey) {
        const response = await this.fetch('/is-account-registered', {
            accountPublicKey: accountPublicKey.toString(),
        });
        return +(await response.text()) === 1;
    }
    async isAliasRegistered(alias) {
        const response = await this.fetch('/is-alias-registered', { alias });
        return +(await response.text()) === 1;
    }
    async isAliasRegisteredToAccount(accountPublicKey, alias) {
        const response = await this.fetch('/is-alias-registered-to-account', {
            accountPublicKey: accountPublicKey.toString(),
            alias,
        });
        return +(await response.text()) === 1;
    }
    async fetch(path, data) {
        const url = new URL(`${this.baseUrl}${path}`);
        const init = data ? { method: 'POST', body: JSON.stringify(data) } : undefined;
        const response = await (0, iso_fetch_1.fetch)(url.toString(), init).catch(() => undefined);
        if (!response) {
            throw new Error('Failed to contact rollup provider.');
        }
        if (response.status === 400) {
            const body = await response.json();
            throw new Error(body.error);
        }
        if (response.status !== 200) {
            throw new Error(`Bad response code ${response.status}.`);
        }
        return response;
    }
}
exports.ServerRollupProvider = ServerRollupProvider;
