// @noir-lang/barretenberg/dest/bridge_id/bridge_id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeId = void 0;
const asset_1 = require("../asset");
const bigint_buffer_1 = require("../bigint_buffer");
const bit_config_1 = require("./bit_config");
const bridge_id_config_1 = require("./bridge_id_config");
const randomInt = (to = 2 ** 30 - 1) => Math.floor(Math.random() * (to + 1));
const getNumber = (val, offset, size) => Number((val >> BigInt(offset)) & ((BigInt(1) << BigInt(size)) - BigInt(1)));
class BridgeId {
    constructor(addressId, inputAssetIdA, outputAssetIdA, inputAssetIdB, outputAssetIdB, auxData = 0) {
        this.addressId = addressId;
        this.inputAssetIdA = inputAssetIdA;
        this.outputAssetIdA = outputAssetIdA;
        this.inputAssetIdB = inputAssetIdB;
        this.outputAssetIdB = outputAssetIdB;
        this.auxData = auxData;
        this.bitConfig = new bit_config_1.BitConfig(inputAssetIdB !== undefined, outputAssetIdB !== undefined);
    }
    static random() {
        return new BridgeId(randomInt(), randomInt(), randomInt(), randomInt(), randomInt(), randomInt());
    }
    static fromBigInt(val) {
        const addressId = getNumber(val, bridge_id_config_1.ADDRESS_OFFSET, bridge_id_config_1.ADDRESS_BIT_LEN);
        const inputAssetIdA = getNumber(val, bridge_id_config_1.INPUT_ASSET_ID_A_OFFSET, bridge_id_config_1.INPUT_ASSET_ID_A_LEN);
        const outputAssetIdA = getNumber(val, bridge_id_config_1.OUTPUT_ASSET_ID_A_OFFSET, bridge_id_config_1.OUTPUT_ASSET_ID_A_LEN);
        const inputAssetIdB = getNumber(val, bridge_id_config_1.INPUT_ASSET_ID_B_OFFSET, bridge_id_config_1.INPUT_ASSET_ID_B_LEN);
        const outputAssetIdB = getNumber(val, bridge_id_config_1.OUTPUT_ASSET_ID_B_OFFSET, bridge_id_config_1.OUTPUT_ASSET_ID_B_LEN);
        const auxData = getNumber(val, bridge_id_config_1.AUX_DATA_OFFSET, bridge_id_config_1.AUX_DATA_LEN);
        const bitConfig = bit_config_1.BitConfig.fromBigInt(BigInt(getNumber(val, bridge_id_config_1.BITCONFIG_OFFSET, bridge_id_config_1.BITCONFIG_LEN)));
        if (!bitConfig.secondInputInUse && inputAssetIdB) {
            throw new Error('Inconsistent second input.');
        }
        if (!bitConfig.secondOutputInUse && outputAssetIdB) {
            throw new Error('Inconsistent second output.');
        }
        return new BridgeId(addressId, inputAssetIdA, outputAssetIdA, bitConfig.secondInputInUse ? inputAssetIdB : undefined, bitConfig.secondOutputInUse ? outputAssetIdB : undefined, auxData);
    }
    static fromBuffer(buf) {
        if (buf.length !== 32) {
            throw new Error('Invalid buffer.');
        }
        return BridgeId.fromBigInt((0, bigint_buffer_1.toBigIntBE)(buf));
    }
    static fromString(str) {
        return BridgeId.fromBuffer(Buffer.from(str.replace(/^0x/i, ''), 'hex'));
    }
    get firstInputVirtual() {
        return (0, asset_1.isVirtualAsset)(this.inputAssetIdA);
    }
    get secondInputVirtual() {
        return !!this.inputAssetIdB && (0, asset_1.isVirtualAsset)(this.inputAssetIdB);
    }
    get firstOutputVirtual() {
        return (0, asset_1.isVirtualAsset)(this.outputAssetIdA);
    }
    get secondOutputVirtual() {
        return !!this.outputAssetIdB && (0, asset_1.isVirtualAsset)(this.outputAssetIdB);
    }
    get secondInputInUse() {
        return this.bitConfig.secondInputInUse;
    }
    get secondOutputInUse() {
        return this.bitConfig.secondOutputInUse;
    }
    get numInputAssets() {
        return this.bitConfig.secondInputInUse ? 2 : 1;
    }
    get numOutputAssets() {
        return this.bitConfig.secondOutputInUse ? 2 : 1;
    }
    toBigInt() {
        return (BigInt(this.addressId) +
            (BigInt(this.inputAssetIdA) << BigInt(bridge_id_config_1.INPUT_ASSET_ID_A_OFFSET)) +
            (BigInt(this.inputAssetIdB || 0) << BigInt(bridge_id_config_1.INPUT_ASSET_ID_B_OFFSET)) +
            (BigInt(this.outputAssetIdA) << BigInt(bridge_id_config_1.OUTPUT_ASSET_ID_A_OFFSET)) +
            (BigInt(this.outputAssetIdB || 0) << BigInt(bridge_id_config_1.OUTPUT_ASSET_ID_B_OFFSET)) +
            (this.bitConfig.toBigInt() << BigInt(bridge_id_config_1.BITCONFIG_OFFSET)) +
            (BigInt(this.auxData) << BigInt(bridge_id_config_1.AUX_DATA_OFFSET)));
    }
    toBuffer() {
        return (0, bigint_buffer_1.toBufferBE)(this.toBigInt(), 32);
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
    equals(id) {
        return id.toBuffer().equals(this.toBuffer());
    }
}
exports.BridgeId = BridgeId;
BridgeId.ZERO = new BridgeId(0, 0, 0);
BridgeId.ENCODED_LENGTH_IN_BYTES = 32;
