// @noir-lang/barretenberg/dest/log/debug.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEnabled = exports.enableLogs = exports.setPostDebugLogHook = exports.setPreDebugLogHook = exports.createDebugLogger = void 0;
const tslib_1 = require("tslib");
const debug_1 = (0, tslib_1.__importDefault)(require("debug"));
let preLogHook;
let postLogHook;
function theFunctionThroughWhichAllLogsPass(logger, ...args) {
    if (preLogHook) {
        preLogHook(...args);
    }
    logger(...args);
    if (postLogHook) {
        postLogHook(...args);
    }
}
function createDebugLogger(name) {
    const logger = (0, debug_1.default)(name);
    return (...args) => theFunctionThroughWhichAllLogsPass(logger, ...args);
}
exports.createDebugLogger = createDebugLogger;
function setPreDebugLogHook(fn) {
    preLogHook = fn;
}
exports.setPreDebugLogHook = setPreDebugLogHook;
function setPostDebugLogHook(fn) {
    postLogHook = fn;
}
exports.setPostDebugLogHook = setPostDebugLogHook;
function enableLogs(str) {
    debug_1.default.enable(str);
}
exports.enableLogs = enableLogs;
function isEnabled(str) {
    return debug_1.default.enabled(str);
}
exports.isEnabled = isEnabled;
