// @noir-lang/barretenberg/dest/sleep/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sleep = exports.InterruptableSleep = void 0;
const errors_1 = require("../errors");
class InterruptableSleep {
    constructor() {
        this.interruptResolve = () => { };
        this.interruptPromise = new Promise(resolve => (this.interruptResolve = resolve));
        this.timeouts = [];
    }
    async sleep(ms) {
        let timeout;
        const promise = new Promise(resolve => (timeout = setTimeout(() => resolve(false), ms)));
        this.timeouts.push(timeout);
        const shouldThrow = await Promise.race([promise, this.interruptPromise]);
        clearTimeout(timeout);
        this.timeouts.splice(this.timeouts.indexOf(timeout), 1);
        if (shouldThrow) {
            throw new errors_1.InterruptError('Interrupted.');
        }
    }
    interrupt(sleepShouldThrow = false) {
        this.interruptResolve(sleepShouldThrow);
        this.interruptPromise = new Promise(resolve => (this.interruptResolve = resolve));
    }
}
exports.InterruptableSleep = InterruptableSleep;
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
exports.sleep = sleep;
